File: Den.Dev.Grunt/Den.Dev.Grunt/Core/Modules/HaloInfinite/StatsModule.cs
// <copyright file="StatsModule.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Den.Dev.Grunt is managed by Halo Studios and Microsoft. This wrapper is not endorsed by Halo Studios or Microsoft.
// </copyright>

using System.Threading.Tasks;
using Den.Dev.Grunt.Core.Foundation;
using Den.Dev.Grunt.Endpoints;
using Den.Dev.Grunt.Models;
using Den.Dev.Grunt.Models.HaloInfinite;

namespace Den.Dev.Grunt.Core.Modules.HaloInfinite
{
    /// <summary>
    /// Module for stats-related API operations including match history and service records.
    /// </summary>
    public class StatsModule : ModuleBase
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="StatsModule"/> class.
        /// </summary>
        /// <param name="client">The client instance to use for API requests.</param>
        internal StatsModule(ClientBase client)
            : base(client, HaloCoreEndpoints.StatsOrigin)
        {
        }

        /// <summary>
        /// Gets challenge decks that are available for a player.
        /// </summary>
        /// <include file='../../../APIDocsExamples/HaloInfinite/Stats_GetChallengeDecks.xml' path='example'/>
        /// <param name="player">The player's numeric XUID.</param>
        /// <returns>An instance of ChallengeDecksResponse containing deck information if request was successful. Return value is null otherwise.</returns>
        public async Task<HaloApiResultContainer<ChallengeDecksResponse, RawResponseContainer>> GetChallengeDecks(string player)
        {
            return await this.GetAsync<ChallengeDecksResponse>($"/hi/players/xuid({player})/decks");
        }

        /// <summary>
        /// Gets the summary on number of played matches.
        /// </summary>
        /// <include file='../../../APIDocsExamples/HaloInfinite/Stats_GetMatchCount.xml' path='example'/>
        /// <param name="player">The player's numeric XUID.</param>
        /// <returns>An instance of PlayerMatchCount containing match counts if request was successful. Return value is null otherwise.</returns>
        public async Task<HaloApiResultContainer<PlayerMatchCount, RawResponseContainer>> GetMatchCount(string player)
        {
            return await this.GetAsync<PlayerMatchCount>($"/hi/players/xuid({player})/matches/count");
        }

        /// <summary>
        /// Gets match history for a player.
        /// </summary>
        /// <include file='../../../APIDocsExamples/HaloInfinite/Stats_GetMatchHistory.xml' path='example'/>
        /// <param name="player">The player's numeric XUID.</param>
        /// <param name="start">Start value for the counter, from which data should be returned.</param>
        /// <param name="count">Number of matches to return. Maximum is 25. Going beyond 25 will result in only 25 values being returned.</param>
        /// <param name="type">Type of matches to query.</param>
        /// <returns>An instance of <see cref="MatchHistoryResponse"/> containing match metadata if request was successful. Return value is null otherwise.</returns>
        public async Task<HaloApiResultContainer<MatchHistoryResponse, RawResponseContainer>> GetMatchHistory(string player, int start, int count, MatchType type)
        {
            return await this.GetAsync<MatchHistoryResponse>(
                $"/hi/players/xuid({player})/matches?start={start}&count={count}&type={type}");
        }

        /// <summary>
        /// Gets stats for a specific match.
        /// </summary>
        /// <include file='../../../APIDocsExamples/HaloInfinite/Stats_GetMatchStats.xml' path='example'/>
        /// <param name="matchId">Match ID in GUID format.</param>
        /// <returns>An instance of MatchStats containing match metadata if request was successful. Return value is null otherwise.</returns>
        public async Task<HaloApiResultContainer<MatchStats, RawResponseContainer>> GetMatchStats(string matchId)
        {
            return await this.GetAsync<MatchStats>($"/hi/matches/{matchId}/stats");
        }

        /// <summary>
        /// Get challenge progression associated with a given match.
        /// </summary>
        /// <include file='../../../APIDocsExamples/HaloInfinite/Stats_GetPlayerMatchProgression.xml' path='example'/>
        /// <param name="player">The player's numeric XUID.</param>
        /// <param name="matchId">Match ID in GUID format.</param>
        /// <returns>An instance of MatchProgression containing match challenge progression metadata if request was successful. Return value is null otherwise.</returns>
        public async Task<HaloApiResultContainer<MatchProgression, RawResponseContainer>> GetPlayerMatchProgression(string player, string matchId)
        {
            return await this.GetAsync<MatchProgression>(
                $"/hi/players/xuid({player})/matches/{matchId}/progression");
        }

        /// <summary>
        /// Gets match privacy settings for a given player.
        /// </summary>
        /// <include file='../../../APIDocsExamples/HaloInfinite/Stats_MatchPrivacy.xml' path='example'/>
        /// <param name="player">The player's numeric XUID.</param>
        /// <returns>An instance of MatchesPrivacy containing match privacy metadata if request was successful. Return value is null otherwise.</returns>
        public async Task<HaloApiResultContainer<MatchesPrivacy, RawResponseContainer>?> MatchPrivacy(string player)
        {
            return await this.GetAsync<MatchesPrivacy>($"/hi/players/xuid({player})/matches-privacy");
        }

        /// <summary>
        /// Gets the service record for a player by their XUID.
        /// </summary>
        /// <remarks>By tweaking season IDs, you can obtain season-specific information such as number of matches played, wins, losses, and others.</remarks>
        /// <include file='../../../APIDocsExamples/HaloInfinite/Stats_GetPlayerServiceRecord.xml' path='example'/>
        /// <param name="xuid">The player's numeric XUID. Example value is "2533274855333605".</param>
        /// <param name="mode">Type of games for which to get the service record.</param>
        /// <param name="seasonId">The ID of the season for which additional stats are pulled. Example value is "Seasons/Season7.json".</param>
        /// <returns>If successful, an instance of <see cref="PlayerServiceRecord"/> containing service record information. Otherwise, returns null with additional details about the error.</returns>
        public async Task<HaloApiResultContainer<PlayerServiceRecord, RawResponseContainer>?> GetPlayerServiceRecordByXuid(string xuid, LifecycleMode mode, string seasonId = "")
        {
            var seasonMarker = !string.IsNullOrWhiteSpace(seasonId) ? $"?seasonId={seasonId}" : string.Empty;

            return await this.GetAsync<PlayerServiceRecord>(
                $"/hi/players/xuid({xuid})/{mode}/servicerecord{seasonMarker}");
        }

        /// <summary>
        /// Gets the service record for a player by their gamertag.
        /// </summary>
        /// <remarks>By tweaking season IDs, you can obtain season-specific information such as number of matches played, wins, losses, and others.</remarks>
        /// <include file='../../../APIDocsExamples/HaloInfinite/Stats_GetPlayerServiceRecord.xml' path='example'/>
        /// <param name="gamertag">The player's gamertag. Example value is "BreadKrtek".</param>
        /// <param name="mode">Type of games for which to get the service record.</param>
        /// <param name="seasonId">The ID of the season for which additional stats are pulled. Example value is "Seasons/Season7.json".</param>
        /// <returns>If successful, an instance of <see cref="PlayerServiceRecord"/> containing service record information. Otherwise, returns null with additional details about the error.</returns>
        public async Task<HaloApiResultContainer<PlayerServiceRecord, RawResponseContainer>?> GetPlayerServiceRecordByGamertag(string gamertag, LifecycleMode mode, string seasonId = "")
        {
            var seasonMarker = !string.IsNullOrWhiteSpace(seasonId) ? $"?seasonId={seasonId}" : string.Empty;

            return await this.GetAsync<PlayerServiceRecord>(
                $"/hi/players/{gamertag}/{mode}/servicerecord{seasonMarker}");
        }

        /// <summary>
        /// Gets the daily custom experience for a player.
        /// </summary>
        /// <include file='../../../APIDocsExamples/HaloInfinite/Stats_GetPlayerDailyCustomExperience.xml' path='example'/>
        /// <param name="player">The player's numeric XUID.</param>
        /// <returns>An instance of <see cref="PlayerDailyCustomExperience"/> containing daily experience if request was successful. Return value is null otherwise.</returns>
        public async Task<HaloApiResultContainer<PlayerDailyCustomExperience, RawResponseContainer>> GetPlayerDailyCustomExperience(string player)
        {
            return await this.GetAsync<PlayerDailyCustomExperience>($"/hi/players/xuid({player})/customexperience");
        }
    }
}

File: src/node/src/modules/halo-infinite/stats.module.ts
import { ModuleBase } from '../base/module-base';
import type { ClientBase } from '../../clients/base/client-base';
import type { HaloApiResult } from '../../models/common/api-result';
import { HALO_CORE_ENDPOINTS } from '../../endpoints/halo-core-endpoints';
import type { MatchType } from '../../models/halo-infinite/enums/match-type';
import type { LifecycleMode } from '../../models/halo-infinite/enums/lifecycle-mode';
import type {
  MatchStats,
  MatchHistoryResponse,
  PlayerMatchCount,
  PlayerServiceRecord,
} from '../../models/halo-infinite/match';
import type { ChallengeDecksResponse, MatchProgression } from '../../models/halo-infinite/progression';
import type { MatchesPrivacy, PlayerDailyCustomExperience } from '../../models/halo-infinite/misc';

/**
 * Stats module for match history and service records.
 *
 * Provides access to:
 * - Match history for players
 * - Individual match statistics
 * - Player service records (career stats)
 * - Challenge decks and progression
 *
 * @example
 * ```typescript
 * // Get match history
 * const history = await client.stats.getMatchHistory('2533274855333605', 0, 25, MatchType.All);
 *
 * // Get specific match details
 * const match = await client.stats.getMatchStats('match-guid-here');
 *
 * // Get player service record
 * const record = await client.stats.getPlayerServiceRecordByXuid('2533274855333605', LifecycleMode.Matchmade);
 * ```
 */
export class StatsModule extends ModuleBase {
  constructor(client: ClientBase) {
    super(client, HALO_CORE_ENDPOINTS.STATS_ORIGIN);
  }

  /**
   * Get challenge decks available for a player.
   *
   * @param player - Player's numeric XUID
   * @returns Challenge decks response
   */
  getChallengeDecks(player: string): Promise<HaloApiResult<ChallengeDecksResponse>> {
    this.assertNotEmpty(player, 'player');
    return this.get<ChallengeDecksResponse>(`/hi/players/xuid(${player})/decks`);
  }

  /**
   * Get match count summary for a player.
   *
   * @param player - Player's numeric XUID
   * @returns Match count breakdown by type
   */
  getMatchCount(player: string): Promise<HaloApiResult<PlayerMatchCount>> {
    this.assertNotEmpty(player, 'player');
    return this.get<PlayerMatchCount>(`/hi/players/xuid(${player})/matches/count`);
  }

  /**
   * Get match history for a player.
   *
   * @param player - Player's numeric XUID
   * @param start - Starting index for pagination (0-based)
   * @param count - Number of matches to return (max 25)
   * @param type - Type of matches to query
   * @returns Paginated match history
   */
  getMatchHistory(
    player: string,
    start: number,
    count: number,
    type: MatchType
  ): Promise<HaloApiResult<MatchHistoryResponse>> {
    this.assertNotEmpty(player, 'player');
    this.assertRange(count, 1, 25, 'count');
    this.assertRange(start, 0, Number.MAX_SAFE_INTEGER, 'start');

    return this.get<MatchHistoryResponse>(
      `/hi/players/xuid(${player})/matches?start=${start}&count=${count}&type=${type}`
    );
  }

  /**
   * Get detailed statistics for a specific match.
   *
   * @param matchId - Match ID in GUID format
   * @returns Complete match statistics
   */
  getMatchStats(matchId: string): Promise<HaloApiResult<MatchStats>> {
    this.assertNotEmpty(matchId, 'matchId');
    return this.get<MatchStats>(`/hi/matches/${matchId}/stats`);
  }

  /**
   * Get challenge progression for a player in a specific match.
   *
   * @param player - Player's numeric XUID
   * @param matchId - Match ID in GUID format
   * @returns Match progression details
   */
  getPlayerMatchProgression(
    player: string,
    matchId: string
  ): Promise<HaloApiResult<MatchProgression>> {
    this.assertNotEmpty(player, 'player');
    this.assertNotEmpty(matchId, 'matchId');

    return this.get<MatchProgression>(
      `/hi/players/xuid(${player})/matches/${matchId}/progression`
    );
  }

  /**
   * Get match privacy settings for a player.
   *
   * @param player - Player's numeric XUID
   * @returns Privacy settings
   */
  getMatchPrivacy(player: string): Promise<HaloApiResult<MatchesPrivacy>> {
    this.assertNotEmpty(player, 'player');
    return this.get<MatchesPrivacy>(`/hi/players/xuid(${player})/matches-privacy`);
  }

  /**
   * Get service record for a player by XUID.
   *
   * The service record contains aggregate career statistics.
   *
   * @param xuid - Player's numeric XUID
   * @param mode - Lifecycle mode (matchmade, custom, local)
   * @param seasonId - Optional season ID for season-specific stats
   * @returns Player service record
   */
  getPlayerServiceRecordByXuid(
    xuid: string,
    mode: LifecycleMode,
    seasonId?: string
  ): Promise<HaloApiResult<PlayerServiceRecord>> {
    this.assertNotEmpty(xuid, 'xuid');

    const seasonParam = seasonId ? `?seasonId=${seasonId}` : '';
    return this.get<PlayerServiceRecord>(
      `/hi/players/xuid(${xuid})/${mode}/servicerecord${seasonParam}`
    );
  }

  /**
   * Get service record for a player by gamertag.
   *
   * @param gamertag - Player's gamertag
   * @param mode - Lifecycle mode (matchmade, custom, local)
   * @param seasonId - Optional season ID for season-specific stats
   * @returns Player service record
   */
  getPlayerServiceRecordByGamertag(
    gamertag: string,
    mode: LifecycleMode,
    seasonId?: string
  ): Promise<HaloApiResult<PlayerServiceRecord>> {
    this.assertNotEmpty(gamertag, 'gamertag');

    const encodedGamertag = encodeURIComponent(gamertag);
    const seasonParam = seasonId ? `?seasonId=${seasonId}` : '';
    return this.get<PlayerServiceRecord>(
      `/hi/players/${encodedGamertag}/${mode}/servicerecord${seasonParam}`
    );
  }

  /**
   * Get daily custom game XP for a player.
   *
   * @param player - Player's numeric XUID
   * @returns Daily custom experience info
   */
  getPlayerDailyCustomExperience(
    player: string
  ): Promise<HaloApiResult<PlayerDailyCustomExperience>> {
    this.assertNotEmpty(player, 'player');
    return this.get<PlayerDailyCustomExperience>(
      `/hi/players/xuid(${player})/customexperience`
    );
  }
}
